Implemented feature introspector to output server feature definitions

KeenanMcCarthy opened the pull request:
In the event of a dump, the feature introspector will iterate over all server features and output
the feature name, as well as its visibility, version, simple name, and symbolic name. 
for #20781

KeenanMcCarthy commented:
#build

tjwatson commented:
#build

tjwatson commented on dev/com.ibm.ws.kernel.feature.core/src/com/ibm/ws/kernel/feature/internal/FeatureIntrospector.java:
```diff
@@ -0,0 +1,67 @@
+        for (String feature : features) {
+            FeatureDefinition featureDef = this.provisioner.getFeatureDefinition(feature);
+            out.println(featureDef.getFeatureName() + "[");
+            out.println("\tvisibility: " + featureDef.getVisibility());
```
I think we should use something like 4 spaces here instead tabs.  I only say this because that seems to be what other introspectors do.

tjwatson commented on dev/com.ibm.ws.kernel.feature.core/src/com/ibm/ws/kernel/feature/internal/FeatureIntrospector.java:
```diff
@@ -0,0 +1,67 @@
+            out.println("\tversion: " + featureDef.getVersion());
+            out.println("\tsimple name: " + featureDef.getFeatureName());
+            out.println("\tsymbolic name: " + featureDef.getSymbolicName());
+            out.println("]");
```
We should be able to safely cast the `featureDef` to `com.ibm.ws.kernel.feature.provisioning.ProvisioningFeatureDefinition` here.  I suggest we do an `instanceof` check and then cast and also provide the following information:

ProvisioningFeatureDefinition.getBundleRepositoryType()
ProvisioningFeatureDefinition.getConstituents(SubsystemContentType)
ProvisioningFeatureDefinition.getSupersededBy()
ProvisioningFeatureDefinition.isAutoFeature()
ProvisioningFeatureDefinition.isSingleton()

KeenanMcCarthy commented:
#build

KeenanMcCarthy commented:
#build

tjwatson commented on dev/com.ibm.ws.kernel.feature.core/src/com/ibm/ws/kernel/feature/internal/FeatureIntrospector.java:
```diff
@@ -0,0 +1,94 @@
+                out.println("    Singleton: " + proFeatDef.isSingleton());
+                out.print("    Superseded By: ");
+                try {
+                    out.println(proFeatDef.getSupersededBy()); //Throws Exception if called outside of provisioning operation
```
I think we can work around this by using internals of this project/bundle.  First you will need to move this class into the package `com.ibm.ws.kernel.feature.internal.subsystem` so that you can use package private methods of that package.  Then you can do something like how `com.ibm.ws.kernel.feature.internal.subsystem.FeatureRepository.LibertyFeatureServiceFactory.getService(Bundle, ServiceRegistration<LibertyFeature>)` does to set the details so we can get this information.

Commit:
Implemented FeatureIntrospector to output server feature definitions

In the event of a dump, the FeatureIntrospector will iterate over all server features and output the feature name, as well as its visibility, version, simple name, and symbolic name. If the feature is a provisioning feature, then the bundle repository type, auto feature, and singleton status will be output. Whether or not the feature is superseded, and what it is superseded by will be output, as well as its constituents and thier associated attributes.

KeenanMcCarthy commented:
#build

tjwatson commented on dev/com.ibm.ws.kernel.feature.core/src/com/ibm/ws/kernel/feature/internal/subsystem/FeatureIntrospector.java:
```diff
@@ -0,0 +1,113 @@
+                    Collection<FeatureResource> constituents = details.getConstituents(null);
+                    out.println("    Constituents: ");
+                    for (FeatureResource constituent : constituents) {
+                        out.println("        " + constituent.getSymbolicName());
```
I would like to see the tolerates information for each included FeatureResource.  Something like this.
```suggestion
                        List<String> tolerates = constituent.getTolerates();
                        String toleratesString = tolerates == null ? "" : ": tolerates:=" + tolerates;
                        out.println("        " + constituent.getSymbolicName() + toleratesString);
```

tjwatson commented on dev/com.ibm.ws.kernel.feature.core/src/com/ibm/ws/kernel/feature/internal/subsystem/FeatureIntrospector.java:
```diff
@@ -0,0 +1,113 @@
+                        out.println("        " + constituent.getSymbolicName());
+                        out.println("            " + constituent.getAttributes().toString());
+                    }
+                    if (detailsAreSet) {
```
Just after the initialization of `detailsAreSet` at line 82 I would add a try-finally block.  Then put this `if (detailsAreSet) {}` code inside the `finally` block.

tjwatson reviewed (commented):
I would be fine merging this as is.  But the build failed in weird ways.  I doubt it failed because of your change.  But since you have to run another build anyways we might as well get these suggestions in first.

Commit:
Updated FeatureIntrospector to output tolerates information for each FeatureResource

If a FeatureResource has a tolerates string, that string is written to output, otherwise no additional information is output. Also moved the provisioning details conditional reset to a finally block, such that it is executed in the event of an exception.

KeenanMcCarthy commented:
Fixes #20781

KeenanMcCarthy commented:
#build

tjwatson commented:
#build

tjwatson commented:
There is a good build in #20998 with the same changes: https://wasrtc.hursley.ibm.com:9443/jazz/resource/itemOid/com.ibm.team.build.BuildResult/_jtbk0MshEeyhmZ5zGPuLaA

The build issues have to do with the size of Keenan's github ID.  We can merge this change.

tjwatson commented:
#libby
